writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
-	if uint64(len(ca.countdown)) < start+1 {
-		return nil, errors.New("count is incorrect length")
-	}
-	for i, num := range ca.countdown {
-		if num == 0 {
-			ca.countdown = ca.countdown[:i+1]
-			break
+
+	// Check if the checksums match the data
+	numbers := make([]byte, numSilos*int64(numIncrease)*4)
+	var cs checksum
+	for _, silo := range silos {
+		// Adjust the size of numbers
+		numbers = numbers[:4*len(silo.numbers)]
+
+		// Read numbers and checksum
+		if _, err := silo.f.ReadAt(numbers, silo.offset); err != nil {
+			return build.ExtendErr("Failed to read numbers of silo", err)
 		}
-		if num != start-uint64(i) {
-			return nil, fmt.Errorf("count is incorrect representation %v != (%v - %v)", num, start, uint64(i))
+		if _, err := silo.f.ReadAt(cs[:], silo.offset+int64(4*len(silo.numbers))); err != nil {
+			return build.ExtendErr("Failed to read checksum of silo", err)
 		}
-	}
 
-	// Check that the splotch file follows the splotch rules.
-	splotchData, err := ioutil.ReadAll(ca.splotchFile)
-	if err != nil {
-		return nil, err
-	}
-	skip := 0
-	for i := 0; i < len(splotchData); i += skip {
-		skip += 80
-		current := binary.LittleEndian.Uint64(splotchData[i : i+8])
-		if current == 0 {
-			continue
-		}
-		for j := 8; j < skip; j += 8 {
-			next := binary.LittleEndian.Uint64(splotchData[j+i : j+i+8])
-			if next != current-1 {
-				return nil, fmt.Errorf("splotch does not count down correctly: %v != (%v - %v)", current, next, 1)
-			}
-			current = next
+		// The checksum should match
+		c := blake2b.Sum256(numbers)
+		if bytes.Compare(c[:checksumSize], cs[:]) != 0 {
+			return errors.New("Checksums don't match")
 		}
 	}
 
-	// Signal that the recovery is complete
 	if err := wal.RecoveryComplete(); err != nil {
-		return nil, err
+		return build.ExtendErr("Failed to signal completed recovery: %v", err)
 	}
-	return ca, nil
+
+	// Close the wal
+	if err := wal.Close(); err != nil {
+		return build.ExtendErr("Failed to close WAL", err)
+	}
+	return nil
 }
 
-// TestWALIntegration creates a plausable use case for the WAL and then
-// attempts to utilize all functions of the WAL.
-func TestWALIntegration(t *testing.T) {
+// TestSilo is an integration test that is supposed to test all the features of
+// the WAL in a single testcase. It uses 100 silos updating 1000 times each.
+func TestSilo(t *testing.T) {
 	if testing.Short() {
 		t.SkipNow()
 	}
 
-	// Create the countdown and extend the count out to 250, closing and
-	// re-opening the wal each time. The number of WAL pages required by each
-	// 'addCount' will incrase over time, giving good confidence that the WAL
-	// is correctly handling multi-page updates.
-	//
-	// 'newCountdown' will chcek that the file is consistent, and detect that
-	// updates are being applied correctly.
-	for r := 0; r < 16; r++ {
-		// Create a folder to house everything we are working with.
-		dir := build.TempDir("wal", t.Name())
-		err := os.MkdirAll(dir, 0700)
+	deps := newFaultyDiskDependency(1000)
```

Comment:
You start the write limit at 1000 and then later reset it to 5000.

I was more thinking that we should just set it at something like 25,000. I'm curious what percentage of the time it hits the writeLimit, but I would like to see it get to the higher numbers every now and then.

Reply:
I think we don't want to set it too high for the first part of the test. Otherwise we have to wait longer for the threads to actually finish since the disk becomes less likely to fail with every write.

Reply:
But you are right. It should say 5000 both times and for the recovery part it should be `math.MaxUint64`